addons/web_editor/static/src/components/upload_progress_toast/upload_service.js

```diff
@@ -67,6 +67,13 @@ export const uploadService = {
              * @param {Function} onUploaded
              */
             uploadFiles: async (files, {resModel, resId, isImage}, onUploaded) => {
```

Comment:
Ok :+1:

Reply:
> Another option would be to check if we can not add an attribute on the `input[type="file"]` to disable natively the selection of folders. There is the `multiple` and `webkitdirectory` attribute. If you can exclude the folders from the selection, it would be 10 times better than letting people pick directory and raise an error 😄 So, in `file_selector.xml`, try setting those attributes and check if you can select a directory from your file explorer.

@julien-banken I have tried every thing with attributes `webkitdirectory` and `multiple` but it did not works for excluding folders.I have tried this in jsfiddle(https://jsfiddle.net/VishalPadhiyar/vfrLatdy/10/) and found that it is behaviour of ubuntu which let you select folder in `input[type='file']`.I have also notice that if you double click on folder after that it won't let you select folder in our case.

Reply:
ok, you can check if the issue is not a simple limitation of the system. If it's not a limitation, you can try the solution I suggested in the other comment? :smile:

Reply:
IMO, it's a system limitation. Upon checking on Windows, I observed that we are unable to select a folder, whereas on Ubuntu, selecting a folder is possible.

Reply:
@julien-banken What to do in this pr ? As it is a limitation of ubuntu so what can we do here ?

Reply:
I think we need to understand what went wrong when uploading multiple files. I can not really help you much as I can not reproduce the issue and I don't have a full traceback that can help me understand what went wrong in the code.

The solution with `multiSelect` got rejected because it involves doing two extra steps in the flow to insert a file with the `/file` command.

Do you think limiting the upload to a single file is the best and only fix? If there is no better solution, I think we can create a variant of `FileSelector` for our modal and pass the first file of the filelist to the file upload service.

Reply:
>I can not reproduce the issue and I don't have a full traceback that can help me understand what went wrong in the code.

![image](https://github.com/odoo/odoo/assets/98956978/5be6791b-70d2-46fa-95d0-a3f90fb6952c)
 - I got this traceback in `master` but i am fixing the issue from `16.0` as we uploading folder doesn't do anything from `16.0.`
![image](https://github.com/odoo/odoo/assets/98956978/712def34-a521-49f0-ae4b-24f9c42a41bc)
- This is the traceback which i got in console:


> Do you think limiting the upload to a single file is the best and only fix?

I do think limiting the upload to single file is best fix but in ubuntu uploading folder doesn't give us a file list.It will simply give folder as a one single file..

> I think we can create a variant of FileSelector for our modal and pass the first file of the filelist to the file upload service.

We don't have filelist when selecting folder using `/file`.We get whole folder as a single file.

Reply:
Coming back to this one :runner:

I read a bunch of stuff on the internet and apparently, you can indeed get a folder from the given filelist. I was wrong :sweat_smile:

On StackOverflow, people suggest trying to read the first bytes of the file you get and check if you get any error when doing so. If you get an error, then the filelist entry is **probably** a folder. Otherwise, the filelist entry is **probably** a file. There does not seem to be a reliable way to distinguish the two.

Source: https://stackoverflow.com/questions/8856628/detecting-folders-directories-in-javascript-filelist-objects

If you can reproduce the issue elsewhere, then we should maybe update the upload service and filter out the entries that are directories. Otherwise, we should scope the changes to our dialog as said in the other comment :smile:

Reply:
Hello @julien-banken ,
``` /**
     * Gets dataURL (base64 data) from the given file or blob.
     * Technically wraps FileReader.readAsDataURL in Promise.
     *
     * @param {Blob|File} file
     * @returns {Promise} resolved with the dataURL, or rejected if the file is
     *  empty or if an error occurs.
     */
    getDataURLFromFile: function (file) {
        if (!file) {
            return Promise.reject();
        }
        return new Promise(function (resolve, reject) {
            var reader = new FileReader();
            reader.addEventListener('load', function () {
                // Handle Chrome bug that creates invalid data URLs for empty files
                if (reader.result === "data:") {
                    resolve(`data:${file.type};base64,`);
                } else {
                    resolve(reader.result);
                }
            });
            reader.addEventListener('abort', reject);
            reader.addEventListener('error', reject);
            reader.readAsDataURL(file);
        });
    },
```

This is the function you are talking about and this is the function from which I get a traceback when uploading a folder.

Reply:
Re-thinking about it :brain: : We are maybe overthinking... We are trying to gracefully manage the upload of files by ignoring folders. But, there is no perfect heuristic to detect that a file is actually a folder in Javascript. And even if we have one, we would have to write more code to transparently manage the error.

We could just let the system crash and show a clear error message to the user to ask them to upload a file. I think it should be more inline with the Odoo policy to do than writing tones of code to gracefully manage all possible errors.

So, here, we could add a simple try/catch when calling that function and open an error dialog with a custom error message. People will just have to click on "ok" and retry the upload. The crash will occur before the server upload so it should not be a problem.

What do you think ?

Reply:
Yes, we can do this.

Reply:
You can do what is easier for you :smile: I think the upload service already report error using a toaster so I guess it's okay to do that.